Providing --path option to transfer-ownership

MorrisJobke opened the pull request:
* downstream of https://github.com/owncloud/core/pull/27343

Tested and works fine here 👍

sharidas committed:
Providing --path option to transfer-ownership

This will help user to selectively move the folders
specified using --path option, instead of moving
entire folder under files directory.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

Update the integration test for transfer-ownership

Update the integration test for transfer-ownership
as the new option --path is introduced in the command.

Signed-off-by: Sujith H <sharidasan@owncloud.com>
Signed-off-by: Morris Jobke <hey@morrisjobke.de>

icewind1991 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -121,6 +133,16 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		Filesystem::initMountPoints($this->sourceUser);
 		Filesystem::initMountPoints($this->destinationUser);
 
+		if (strlen($this->inputPath) >= 1) {
```
`if($this->inputPath)` would be better, otherwise single character paths can lead to unexpected results

MorrisJobke replied:
I don't get this. It checks if the path is greater equal than.

icewind1991 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -155,7 +177,14 @@ protected function analyse(OutputInterface $output) {
 		$self = $this;
-		$this->walkFiles($view, "$this->sourceUser/files",
+		$walkPath = "$this->sourceUser/files";
+		if ( strlen($this->inputPath) > 0) {
```
just `if($this->inputPath)` would be cleaner

nickvergessen replied:
but that fails directory `'0'`

icewind1991 replied:
`===''` then

icewind1991 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -88,6 +92,12 @@ protected function configure() {
+				'path',
+				null,
+				InputOption::VALUE_REQUIRED,
+				'selectively provide the path to transfer. For example --path="folder_name"'
```
It seems to expect absolute path `/$user/files/path`, the cli help should reflect that

MorrisJobke replied:
Not really - see line 139.

icewind1991 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -155,7 +177,14 @@ protected function analyse(OutputInterface $output) {
-		$this->walkFiles($view, "$this->sourceUser/files",
+		$walkPath = "$this->sourceUser/files";
+		if ( strlen($this->inputPath) > 0) {
+			if ($this->inputPath !== "$this->sourceUser/files") {
```
This `if` only makes the code more complex imo

icewind1991 commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -216,9 +245,20 @@ private function collectUsersShares(OutputInterface $output) {
-		$view->rename("$this->sourceUser/files", $this->finalTarget);
-		// because the files folder is moved away we need to recreate it
-		$view->mkdir("$this->sourceUser/files");
+		$sourcePath = (strlen($this->inputPath) > 0) ? $this->inputPath : "$this->sourceUser/files";
```
Simply setting `$this->inputPath` to a default value of `$this->sourceUser/files` would remove the need for endless if statements like this one

MorrisJobke committed:
improve file transfer code

Signed-off-by: Morris Jobke <hey@morrisjobke.de>

MorrisJobke commented:
@icewind1991 I addressed all the comments you had and refactored out most of the if statements.

I tested it with and without the path option and it works fine 👍

icewind1991 reviewed (approved):
Much better
